source/common/config/unified_mux/delta_subscription_state.cc

```diff
@@ -0,0 +1,197 @@
+  absl::flat_hash_set<std::string> names_added_removed;
+  Protobuf::RepeatedPtrField<envoy::service::discovery::v3::Resource> non_heartbeat_resources;
+  for (const auto& resource : message.resources()) {
+    if (!names_added_removed.insert(resource.name()).second) {
+      throw EnvoyException(
+          fmt::format("duplicate name {} found among added/updated resources", resource.name()));
+    }
+    if (isHeartbeatResource(resource)) {
+      continue;
+    }
+    // TODO (dmitri-d) consider changing onConfigUpdate callback interface to avoid copying of
+    // resources
+    non_heartbeat_resources.Add()->CopyFrom(resource);
+    // DeltaDiscoveryResponses for unresolved aliases don't contain an actual resource
+    if (!resource.has_resource() && resource.aliases_size() > 0) {
+      continue;
+    }
+    if (message.type_url() != resource.resource().type_url()) {
+      throw EnvoyException(fmt::format("type URL {} embedded in an individual Any does not match "
+                                       "the message-wide type URL {} in DeltaDiscoveryResponse {}",
+                                       resource.resource().type_url(), message.type_url(),
+                                       message.DebugString()));
+    }
+  }
+  for (const auto& name : message.removed_resources()) {
+    if (!names_added_removed.insert(name).second) {
+      throw EnvoyException(
+          fmt::format("duplicate name {} found in the union of added+removed resources", name));
+    }
+  }
+
+  {
+    const auto scoped_update = ttl_.scopedTtlUpdate();
+    for (const auto& resource : message.resources()) {
+      addResourceState(resource);
+    }
+  }
+
+  callbacks().onConfigUpdate(non_heartbeat_resources, message.removed_resources(),
+                             message.system_version_info());
+
+  // If a resource is gone, there is no longer a meaningful version for it that makes sense to
+  // provide to the server upon stream reconnect: either it will continue to not exist, in which
+  // case saying nothing is fine, or the server will bring back something new, which we should
+  // receive regardless (which is the logic that not specifying a version will get you).
+  //
+  // So, leave the version map entry present but blank. It will be left out of
+  // initial_resource_versions messages, but will remind us to explicitly tell the server "I'm
+  // cancelling my subscription" when we lose interest.
+  for (const auto& resource_name : message.removed_resources()) {
+    if (resource_state_.find(resource_name) != resource_state_.end()) {
+      resource_state_[resource_name] = ResourceState::waitingForServer();
+    }
+  }
+  ENVOY_LOG(debug, "Delta config for {} accepted with {} resources added, {} removed", typeUrl(),
+            message.resources().size(), message.removed_resources().size());
+}
+
+std::unique_ptr<envoy::service::discovery::v3::DeltaDiscoveryRequest>
+DeltaSubscriptionState::getNextRequestInternal() {
```

Comment:
The original had a node merge at the end, this isn't here?

Reply:
This bit has been moved to the mux: https://github.com/envoyproxy/envoy/pull/17049/commits/109ac32f8661ffed4dad2f607c6eec28b221435a#diff-fe56236f0649ba734162e4fa294465a6004c393242c0af89421cbd5fe279fe58R338. Currently a node is always being added in delta requests. With this change both delta and sotw send a node if: this is a first request in the stream, skip_subsequent_node is false, or dynamic context has changed. 

I think I should change `MergeFrom` to `CopyFrom` when copying node, ro make sure we never end up with multiple copies of it.

Reply:
Ack, makes sense.